### PR TITLE
perforce: Perform changelist mapping for new and existing depots imported as git repos

### PR DESCRIPTION
**Pull request by indradhanush**
Part of #40330.

## Why

For a perforce depot, we want to be able to answer the question: "What is the changelist ID when this change was made?". In a previous PR (#51872) we added the DB layer to support this (`repo_commits_changelists` table). And in this PR we add the job management pipeline and the job execution itself.

## What

A new "plugin" / "service" named `Perforce` is added to gitserver.Server. It is responsible for all things perforce and exists as a sub-package of `cmd/gitserver/server`. This makes it easy to break out of gitserver if needed in the future and also allows to be tested independently of gitserver.

For every perforce depot, the job management pipeline will:

1. For a new repo, map all the commits to their corresponding changelist IDs in the DB
2. For existing repos, when a repo is synced all commits will be mapped to their corresponding changelist IDs in the DB
3. For new commits (changelists) synced to a repo, the newly discovered commits will be mapped to the DB

## Caveats

1. If a repo is deleted from the DB, we will drop all rows from the `repo_commits_changelists` table at the moment
2. If a repo is "deleted" from the code host, we do not touch the entries in the `repo_commits_changelists` table. This means if a repo is corrupted and users want to sync a repo from a clean slate, we won't be overwriting the existing rows. This needs to be handled gracefully and will follow up in a subsequent PR
3. If a repo has a force push, the last mapped commit in the DB may no longer be a valid revision. This will mean the `lastMappedCommit..HEAD` range will no longer be valid. This needs to be handled gracefully and will follow up in a subsequent PR

## Test plan

1. Added tests
2. Tested locally with `sg start` for existing repos, newly synced repos and existing repos with new commits
3. Builds should pass

### CODE REVIEW

File: cmd/gitserver/server/server.go [hunk 1]
```diff
@@ -2405,6 +2407,8 @@ func (s *Server) doClone(ctx context.Context, repo api.RepoName, dir GitDir, syn
 	logger.Info("repo cloned")
 	repoClonedCounter.Inc()
 
+	s.PerforceChangelistMappingQueue.push(&perforceChangelistMappingJob{repo: repo})
```
Comment (indradhanush) [inline cmd/gitserver/server/server.go:2410]: TODO: Check if repo is perforce depot before pushing.

File: cmd/gitserver/server/perforce.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+var gitP4Pattern = lazyregexp.New(`\[(?:git-p4|p4-fusion): depot-paths = "(.*?)"\: change = (\d+)\]`)
+
+func getP4ChangelistID(body string) (string, error) {
+	matches := gitP4Pattern.FindStringSubmatch(body)
+	if len(matches) != 3 {
+		return "", errors.Newf("failed to retrieve changelist ID from commit body: %q", body)
+	}
+
+	return matches[2], nil
+}
```
Comment (indradhanush) [inline cmd/gitserver/server/perforce.go:245]: TODO: Duplicated code from GQL layer. Move to `internal/perforce` and use in both places.

File: cmd/gitserver/server/perforce.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+	repo api.RepoName
+}
+
+type perforceChangelistMappingQueue struct {
+	mu   sync.Mutex
+	jobs *list.List
+
+	cmu  sync.Mutex
+	cond *sync.Cond
+}
```
Comment (indradhanush) [inline cmd/gitserver/server/perforce.go:31]: TODO: Pretty much a copy paste of CloneQueue to get this to start working. Use generics as a follow up.

File: cmd/gitserver/server/perforce.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+		return "", &GitCommandError{Err: err, Output: string(output)}
+	}
+
+	return string(output), nil
+}
+
+func newMappableCommits(ctx context.Context, logger log.Logger, dir GitDir, lastMappedCommit, head string) ([]types.PerforceChangelist, error) {
+	cmd := exec.CommandContext(ctx, "git", "log")
+	if lastMappedCommit != "" {
+		cmd.Args = append(cmd.Args, fmt.Sprintf("%s..%s", lastMappedCommit, head))
```
Comment (keegancsmith) [inline cmd/gitserver/server/perforce.go:209]: You need to handle lastMappedCommit..head failing. eg
- gitserver disk being restored from backup meaning head is an ancestor of lastMappedCommit.
- lastMappedCommit no longer existing due to re-import from perforce. In this case you need to handle the sha <-> changelist mapping being invalid.

File: cmd/frontend/graphqlbackend/perforce_changelist.go [hunk 5]
```diff
@@ -43,7 +43,9 @@ func parseP4FusionCommitSubject(subject string) (string, error) {
 	return matches[2], nil
 }
 
-// Either git-p4 or p4-fusion could be used to convert a perforce depot to a git repo. In which case the
+// Either git-p4 or p4-fusion could have been used to convert a perforce depot to a git repo. In
+// which case the commit message would look like:
+//
 // [git-p4: depot-paths = "//test-perms/": change = 83725]
 // [p4-fusion: depot-paths = "//test-perms/": change = 80972]
 var gitP4Pattern = lazyregexp.New(`\[(?:git-p4|p4-fusion): depot-paths = "(.*?)"\: change = (\d+)\]`)
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/perforce_changelist.go:51]: out of interest do you know why git-p4 and p4-fusion use this format of trailers in git commit messages? Git has an official trailer format which integrates into commands. You've probably seen it before with things like co-authored-by: foo. If you are interested the docs are in PRETTY FORMATS section of git log, and `git help interpret-trailers`

Comment (keegancsmith) [reply cmd/frontend/graphqlbackend/perforce_changelist.go:51]: FYI this regex is likely overly restrictive. Here is the code in git-p4 which actually parses these lines. I would imitate that, even though it doesn't properly support parsing quoted strings. https://github.com/git/git/blob/5bc069e383539824fd3a0d897100d44bbe1f8a24/git-p4.py#L1018-L1041

File: cmd/gitserver/server/perforce.go (hunk 4 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce.go:209]: Thanks! This is what I'm thinking on how I should handle this:

1. Before invoking `newMappableCommits`, check if `git rev-parse <lastMappedCommit` works succesfully
2. If it is succesful, go ahead as above
3. If it fails, start a binary search on the list of commits mapped in the DB to find the `lastMappedCommit` in the DB that also exists in the repo on disk
4. When step 3 above returns with the correct `lastMappedCommit`, drop all rows in the DB following it for this repo
5. Start with step 2 to resume mapping from this point


What do you think?

**Comment by indradhanush** [comment]
> What is the motivation for changelist mapping job living on gitserver? The hint to me that it maybe should live somewhere else is re-implementing logic like parsing HEAD and git log + exporting of logic that lives in the gitserver client for use in the server

I am seeing the changelist mapping job as a side effect of a clone / fetch operation for perforce depots. Initially I started to write this in the client and then it felt like an extra round of back and forth which we didn't need since all the information we need is already available in the clone / fetch stage inside Gitserver.

> The gitserver/server package is already quite large. This code atleast feels like it could live in a subpackage of server if indeed it belongs as part of gitserver's server.

👍

File: cmd/frontend/graphqlbackend/perforce_changelist.go (hunk 5 above)
Comment (indradhanush) [reply cmd/frontend/graphqlbackend/perforce_changelist.go:51]: > out of interest do you know why git-p4 and p4-fusion use this format of trailers in git commit messages? 

I do not know the historical reason behind git-p4 choosing this format, but p4-fusion most likely wanted to stay consistent with git-p4 except the `p4-fusion` string.

File: cmd/gitserver/server/perforce.go (hunk 3 above)
Comment (mrnugget) [reply cmd/gitserver/server/perforce.go:33]: Why use this and not use our `workers`-databased-backed job? I'd say if there is one big lesson from a lot of gitserver stuff over the years it's that in-memory queues are hard to debug and often a source of bugs/confusion.

Comment (indradhanush) [reply cmd/gitserver/server/perforce.go:33]: I think that's a really good point and I agree, just didn't think of it. When I started this implementation I got easily influenced by the existing paradigm of the `CloneQueue` as something that works and the temptation to use this was too strong. :P

Comment (indradhanush) [reply cmd/gitserver/server/perforce.go:33]: To clarify - I'm now looking into implementing this as a worker job now.

Comment (mrnugget) [reply cmd/gitserver/server/perforce.go:33]: Great. @alexAtSourcegraph and @thenamankumar did this for permission syncing, they can give you pointers.

Also, see the docs: https://docs.sourcegraph.com/dev/background-information/workers#adding-a-new-worker

(And take a look at the other workers that are running in `workers` process)

File: cmd/gitserver/server/perforce.go (hunk 4 above)
Comment (keegancsmith) [reply cmd/gitserver/server/perforce.go:219]: binary search is a fun solution here, I like it. Man having sequential revision IDs makes life so much nicer.

**Comment by keegancsmith** [comment]
We have other bits of code in search and code-intel which react to commits coming in. But to be honest that stuff could be "faster" to react, and we want this to be super fast due to UX implications if it isn't. I do think in the future it would make sense for things like this not to be happening on gitserver, but for now moving the code into a sub-package and making minimal changes to the main server pkg itself sounds good to me.

This made me think, we need to make sure all code which consumes this information is aware it can be out of date.IE there will always be a delay (although small) where the changeset won't be mapped yet after a p4-fusion run.

**Comment by indradhanush** [comment]
> This made me think, we need to make sure all code which consumes this information is aware it can be out of date.IE there will always be a delay (although small) where the changeset won't be mapped yet after a p4-fusion run.

Yes, the early plan is that we build out the mapping code and show a 404 for pages that depend on this and the information is not computed yet. In the meantime, we want to introduce a cache as a follow up that will serve as an alternative while the computation is in progress behind the scenes. But getting the functionality working on the ideal path of having the commit <-> changelist map created in the DB and then a user is able to access it is first priority.

File: cmd/gitserver/server/perforce.go (hunk 3 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce.go:33]: @mrnugget I wanted to leave a quick update on next steps in regards to this discussion:

I got on a call with @keegancsmith and @peterguy yesterday and discussed the pros and cons of using the worker vs using the clonequeue implementation and we all agreed that at the moment, the worker is adding unnecessary complexity by forcing us to have to send the git-log output over to the worker to process the job or on the other case do the job in the gitserver but just use the worker as a signaling mechanism.

The clonequeue implementation is well tested and proven to work reliably and given that this mapping job really needs to be done in gitserver there isn't a lot to be gained with adding the worker here. Additionally, we need the mapping job to complete as soon as possible and do not want to be held back in case the worker is backed up by other jobs. The urgency in completing the mapping job is required because without the mapping job completed a repo although cloned has partial functionality in the case of perforce depots.

We made some other design decisions about how we want to be inserting the rows during the first import and during each successive import (TL;DR: We want to insert them all at once or fail so that we don't have partially mapped data in the DB which makes life harder to reconcile for successive runs, see [this Slack thread](https://sourcegraph.slack.com/archives/C018CJQ5YBB/p1684310160823739) where I also asked in #chat-databases for advice on this decision and got some good suggestions), and directly executing the job in gitserver also makes this easier and removes one extra step of inter-service coordination.

If we observe any major difficulties with this approach in testing or in production, it will be easy to migrate the mapping job over to the worker without causing any disruptions to the existing customers / repos.

File: cmd/gitserver/server/perforce.go [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+// e2f6d6e306490831b0fdd908fdbee702d7074a66 [p4-fusion: depot-paths = "//rhia-depot-test/": change = 83732]
+// 90b9b9574517f30810346f0ab07f66c49c77ab0f [p4-fusion: depot-paths = "//rhia-depot-test/": change = 83731]
+var logFormatWithCommitSHAAndCommitBodyOnly = "--format=format:%H %b"
+
+func newMappableCommits(ctx context.Context, logger log.Logger, dir GitDir, lastMappedCommit, head string) ([]types.PerforceChangelist, error) {
+	cmd := exec.CommandContext(ctx, "git", "log")
+	// FIXME: When lastMappedCommit..head is an invalid range.
+	// TODO: Follow up in a separate PR.
+	if lastMappedCommit != "" {
+		cmd.Args = append(cmd.Args, fmt.Sprintf("%s..%s", lastMappedCommit, head))
```
Comment (evict) [inline cmd/gitserver/server/perforce.go:231]: This is dangerous, `git log` supports an argument `--output`. If I manage to inject `--output=../path/outside/dir` I can write files. Although I cannot 100% control the content, it's still dangerous and I'd suggest making sure this cannot happen.

Comment (indradhanush) [reply cmd/gitserver/server/perforce.go:231]: Thanks for the call out! This is great insight that and I wasn't thinking of this angle at all. Overall this should be still pretty difficult to inject since a user cannot fake commit SHAs as far as I know, but I will ensure that we safely handle non commit-SHA like strings when I follow up with another PR. Very useful thing to keep in mind for future "exec" like code that I write. Thank you! 💟

**Comment by indradhanush** [comment]
It seems that I have a flakey test in `TestGetCommitsToInsert/new_repo,_never_mapped` that fails 1 in 500. Investigating this next.

File: cmd/gitserver/server/perforce/perforce.go [hunk 7]
```diff
@@ -216,78 +214,53 @@ var logFormatWithCommitSHAAndCommitBodyOnly = "--format=format:%H %b"
...
+		return nil, err
+	}
 
-		output, err := common.RunWith(ctx, wrexec.Wrap(ctx, logger, cmd), false, progressWriter)
-		if err != nil {
-			return &common.GitCommandError{Err: err, Output: string(output)}
-		}
-		return nil
-	})
+	cmd.Start()
```
Comment (keegancsmith) [inline cmd/gitserver/server/perforce/perforce.go:234]: you need to check for error here, sorry missed that

**Comment by indradhanush** [comment]
Rebased off main to fix merge conflicts (bazel files) and dropped the use of `common.RunWith` in `headCommitSHA` function.

File: cmd/gitserver/server/server.go [hunk 8]
```diff
@@ -1075,6 +1080,8 @@ func (s *Server) handleRepoUpdate(w http.ResponseWriter, r *http.Request) {
 		// An update error "wins" over a status error.
 		if updateErr != nil {
 			resp.Error = updateErr.Error()
+		} else {
+			s.maybeEnqueuePerforceChangelistMappingJob(ctx, logger, req.Repo, dir)
```
Comment (keegancsmith) [inline cmd/gitserver/server/server.go:1084]: High level design: this logic feels like it should live inside the Syncer implementation for perforce. I just checked the interface out and the problem is that CloneCommand returns an exec.Command so you don't have any hooks here. It feels wrong to fix that here, but maybe we should extend the syncer interface to have a post-fetch hook?

That would also help remove some of the tests you wrote ensuring we call this function.

Then a future nice refactor would be working out how we can make the syncer interface less clunky. Fetch does the fetch. Clone returns a command. Some implementations of clone just do the clone then return a noop clone command. IE the interface is broken.

File: cmd/gitserver/server/perforce/perforce.go [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/perforce"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+type PerforceService interface {
+	EnqueueChangelistMappingJob(*ChangelistMappingJob)
+	StartPerforceChangelistMappingPipeline(context.Context)
```
Comment (keegancsmith) [inline cmd/gitserver/server/perforce/perforce.go:26]: preference: you can reduce the API surface and make this harder to mess up. I recommend you remove this method. Instead on `EnqueueChangelistMappingJob` you can have a sync.Once protecting initialization.

File: cmd/gitserver/server/perforce/perforce.go [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+			}
+		}
+	}
+}
+
+// changelistMappingConsumer "consumes" jobs "produced" by the producer.
+func (s *Service) changelistMappingConsumer(ctx context.Context, jobs <-chan *ChangelistMappingJob) {
+	logger := s.Logger.Scoped("changelistMappingConsumer", "process perforce changelist mapping jobs")
+
+	// Process only one job at a time for a simpler pipeline at the moment.
```
Comment (keegancsmith) [inline cmd/gitserver/server/perforce/perforce.go:94]: do we intend on making it possible to have more than 1 job? Right now there is a whole bunch of complicated machinery which could all just be us reading from the queue. IE you introduce a channel and a queue reader and a channel pusher. We could just read from the queue.

Additionally if you ever where to have multiple queue consumers this implementation would still be broken because you need to introduce some sort of concurrency control to prevent multiple writers for the same repo.

Basically, I would recommend just simplifying all of this and just reading from the queue.

File: cmd/gitserver/server/perforce/perforce.go [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+		logger := logger.With(log.String("job.repo", string(j.RepoName)))
+
+		select {
+		case <-ctx.Done():
+			logger.Error("context done", log.Error(ctx.Err()))
+			return
+		default:
+		}
+
+		err := s.doChangelistMapping(ctx, j)
```
Comment (keegancsmith) [inline cmd/gitserver/server/perforce/perforce.go:105]: I would recommend adding some sort of panic protection here. If one of these jobs where to panic it would take down the whole gitserver.

File: cmd/gitserver/server/perforce/perforce.go [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+	queue := common.NewQueue[ChangelistMappingJob](jobs)
+
+	return &Service{
+		Logger:                 logger,
+		DB:                     db,
+		changelistMappingQueue: queue,
+	}
+}
+
+func (s *Service) EnqueueChangelistMappingJob(job *ChangelistMappingJob) {
```
Comment (keegancsmith) [inline cmd/gitserver/server/perforce/perforce.go:51]: it is cheap to check the site configuration. Why don't you just check here if the mapping job is enabled? That way the business logic around mapping is nearly all contained in this file.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
LGTM have a few comments I'd like to follow up on so requesting changes.

Additionally I think you should consider somehow gating this feature behind its own feature flag. There is a possibility this could produce a lot of logspam at a p4 user since:
- we don't yet support sha <-> rev changing. All inserts will just fail if I am not mistaken.
- we have very rigid parsing of p4 commit message bodies. It feels like it should be a little more permissive (ie should match the logic used by git-p4).

File: cmd/gitserver/server/perforce/perforce.go (hunk 10 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce/perforce.go:94]: > do we intend on making it possible to have more than 1 job? 

Yes. I am planning to make it possible to run the same number of concurrent jobs as the number of concurrent clones (reuse the same config too and update the doc) so I used this PR to introduce the support structure for this and was intending to follow up once this is tested with perforce on S2 / Dogfood. Otherwise this would be quite a bit of a bottleneck for customers with heavier perforce workloads.

> Additionally if you ever where to have multiple queue consumers this implementation would still be broken because you need to introduce some sort of concurrency control to prevent multiple writers for the same repo.

Yep! I left out the concurrency limiters here because they weren't being used yet and would have added them in the follow up PR post testing on production. The code would be very similar (borderline copy-paste) to the `acquireclonelimiter` code we already have in place for managing the clone queues.

I agree that without the parallel job execution in-place this does feel like quite a bit of unnecessary job management and would be much simpler to just read from the queue one at a time. With that context in mind, would you still recommend to cut the clutter and simplify this code for now and only introduce this design in the follow up PR? 

My motivation to introduce this here was to have the producer-consumer pipeline tested before I added parallel job exection on it. I'm happy to follow your recommendation on this one either way. 🙂

File: cmd/gitserver/server/perforce/perforce.go (hunk 12 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce/perforce.go:51]: Yes, adding the check for the mapping job being enabled here before pushing onto the queue would make sense (also agreed that we should gate it behind a site config).

For clarification - the current check in the [maybeEnqueuePerforceChangelistMappingJob](https://github.com/sourcegraph/sourcegraph/pull/51860/files/0d78121c86b0b3be4af4dd6f9c3751efbe8a3eb4#diff-f5ac848d014dd8443978e31081a9070426aecc0fc35ab065b1441f61b16452ebR2970-R2980) func is to check if this is a perforce depot and only then push on the queue. Did you also mean to bring that check inside the PerforceService API?

File: cmd/gitserver/server/perforce/perforce.go (hunk 10 above)
Comment (keegancsmith) [reply cmd/gitserver/server/perforce/perforce.go:94]: alright that makes sense to me.

File: cmd/gitserver/server/perforce/perforce.go (hunk 12 above)
Comment (keegancsmith) [reply cmd/gitserver/server/perforce/perforce.go:51]: yes I think you should do that too. My other suggestion mentioned this being part of the Syncer, then you wouldn't need to do that since you would know this was p4 synced.

File: cmd/gitserver/server/perforce/perforce.go (hunk 9 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce/perforce.go:26]: I'm not sure how I feel about this. It feels like it would make it harder to predict how this service works over time (I could be wrong but I'm tempted to avoid the potential pit-falls of using sync.Once in case there's something that I miss). How about I do this as a side effect of calling `perforce.NewService` instead?

Comment (keegancsmith) [reply cmd/gitserver/server/perforce/perforce.go:26]: alright NewService sounds good. A lot of my go preferences in design are around using structs instead of interfaces and making sure the zero value is functional :P But NewService sounds like a good place to start the goroutine.

**Review by keegancsmith — APPROVED** [review]
@indradhanush I know you wanna do a quick sync on this, but I can only zoom later today. Anything I can help respond to to unblock this PR now? Based on the inline comments I don't think there is anything larger left (just some smaller things), so approving :)

File: cmd/gitserver/server/server.go (hunk 8 above)
Comment (indradhanush) [reply cmd/gitserver/server/server.go:1084]: @keegancsmith I started redesigning this yesterday and ran into some thoughtful challenges. Overall I agree that the syncer is the right place for perforce related stuff as a post clone / fetch hook. But with the current `PerforceService` designed as a "plugin" to `gitserver.Server`, to enqueue the job from inside the perforce syncer I'll need to pass the server or the `PerforceService` to the syncer as well. Which is beginning to make it look ugly and then it gets tricky to ensure its set every where the syncer is used. The challenge feels a symptom of the clunky interface design at the moment.

**Comment by indradhanush** [comment]
> @indradhanush I know you wanna do a quick sync on this, but I can only zoom later today. Anything I can help respond to to unblock this PR now? Based on the inline comments I don't think there is anything larger left (just some smaller things), so approving :)

@keegancsmith, Thanks for the approval! [This](https://github.com/sourcegraph/sourcegraph/pull/51860#discussion_r1219085526) comment is the only thing that is kind of a blocker for me for which I thought a zoom call might be better. I'm all set to follow up on the other comments and not blocked on them.

File: cmd/gitserver/server/server.go (hunk 8 above)
Comment (keegancsmith) [reply cmd/gitserver/server/server.go:1084]: yup I agree there are a few smells in the interface. I think stick with what you are currently doing and land this and make forward progress :) Branch cut coming soon, lets get this out and start testing it.

Lets have a call tomorrow morning (I will be at home) and do some pairing on improving the Syncer interface. I think an hour of pairing should make some nice changes.

Comment (indradhanush) [reply cmd/gitserver/server/server.go:1084]: Sounds good. Let's do it tomorrow! (I'll send an invite)

File: cmd/gitserver/server/perforce.go [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+	}
+
+	dir.Set(cmd)
+	output, err := runWith(ctx, wrexec.Wrap(ctx, logger, cmd), false, nil)
+	if err != nil {
+		return nil, &GitCommandError{Err: err, Output: string(output)}
+	}
+
+	commits, err := gitserver.ParseGitLogOutput(output)
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply cmd/gitserver/server/perforce.go]: We use a buffered scanner now so unbounded size from git log output is no longer an issue.

File: cmd/gitserver/server/server.go (hunk 8 above)
Comment (indradhanush) [reply cmd/gitserver/server/server.go]: I've moved this logic inside the `perforce.Service`. Also now that I no longer need to test if this method was invoked, I didn't need the mocks which means I didn't need the `PerforceService` interface either. So I've also removed the interface and directly using the struct in definitions.

File: cmd/gitserver/server/perforce/perforce.go (hunk 9 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce/perforce.go]: `NewService` now kicks off the producer consumer pipeline. Also `StartPerforceChangelistMappingPipeline` is now a private method of the struct. I've added docs on the struct and `NewService` definition that indicate that:

1. Only `NewService` should be used to instantiate `perforce.Service`
2. The side effect of calling `NewService`

File: cmd/gitserver/server/perforce/perforce.go (hunk 12 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce/perforce.go:63]: Updated `EnqueueChangelistMappingJob` so that now it is:

1. Guarded by a new experimental feature config (`perforceChangelistMapping`)
2. Also checks if the repo belongs to a perforce code host

File: cmd/gitserver/server/perforce/perforce.go (hunk 11 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce/perforce.go:128]: @keegancsmith Last ask for the time being: I used `goroutine.Go` for panic protection here: https://github.com/sourcegraph/sourcegraph/pull/51860/commits/10ca879acc7ca13a8c2c2816e403a778628dbc02

Does this look good or should I be trying to restart the pipeline on panic?

File: cmd/gitserver/server/perforce.go (hunk 6 above)
Comment (indradhanush) [reply cmd/gitserver/server/perforce.go]: @evict Follow up PR is here to fix the security concern: https://github.com/sourcegraph/sourcegraph/pull/53468

File: cmd/gitserver/server/perforce/perforce.go [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+	}
+}
+
+func (s *Service) startPerforceChangelistMappingPipeline(ctx context.Context) {
+	jobs := make(chan *ChangelistMappingJob)
+
+	// Protect against panics.
+	goroutine.Go(func() { s.changelistMappingConsumer(ctx, jobs) })
+	goroutine.Go(func() { s.changelistMappingProducer(ctx, jobs) })
+}
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply cmd/gitserver/server/perforce/perforce.go:82]: Done: https://github.com/sourcegraph/sourcegraph/pull/53325
